MPF.Core/Utilities/Tools.cs

```diff
@@ -642,5 +643,409 @@ public static bool ParseGetKeyLog(string? logPath, out byte[]? key, out byte[]?
+                br.BaseStream.Seek(8, SeekOrigin.Begin);
+                return new string(br.ReadChars(8));
+            }
+            catch
+            {
+                return string.Empty;
+            }
+        }
+
+        /// <summary>
+        /// Get the XGD2/3 Master ID (XeMID) information
+        /// </summary>
+        /// <param name="dmi">DMI.bin file location</param>
+        /// <returns>String representation of the XGD2/3 DMI information, empty string on error</returns>
+        public static string GetXGD23XeMID(string dmi)
+        {
+            if (!File.Exists(dmi))
+                return string.Empty;
+
+            try
+            {
+                using var br = new BinaryReader(File.OpenRead(dmi));
+                br.BaseStream.Seek(64, SeekOrigin.Begin);
+                return new string(br.ReadChars(14));
+            }
+            catch
+            {
+                return string.Empty;
+            }
+        }
+
+        /// <summary>
+        /// Get XGD type from SS.bin file
+        /// </summary>
+        /// <param name="ss"></param>
+        /// <param name="xgdType"></param>
+        /// <returns></returns>
+        public static bool GetXGDType(string? ss, out int xgdType)
+        {
+            xgdType = 0;
+
+            if (string.IsNullOrEmpty(ss) || !File.Exists(ss))
+                return false;
+
+            using FileStream fs = File.OpenRead(ss);
+            byte[] buf = new byte[3];
+            int numBytes = fs.Read(buf, 13, 16);
+
+            if (numBytes != 3)
+                return false;
+
+            return GetXGDType(buf, out xgdType);
+        }
+
+        /// <summary>
+        /// Get XGD type from SS.bin sector
+        /// </summary>
+        /// <param name="ss">Byte array of SS.bin sector</param>
+        /// <param name="xgdType">XGD type</param>
+        /// <returns>True if successful, false otherwise</returns>
+        public static bool GetXGDType(byte[] ss, out int xgdType)
+        {
+            xgdType = 0;
+
+            // Concatenate the last three values
+            long lastThree = (((ss[13] << 8) | ss[14]) << 8) | ss[15];
+
+            // Return XGD type based on value
+            switch (lastThree)
+            {
+                case 0x2033AF:
+                    xgdType = 1;
+                    return true;
+                case 0x20339F:
+                    xgdType = 2;
+                    return true;
+                case 0x238E0F:
+                    xgdType = 3;
+                    return true;
+                default:
+                    return false;
+            }
+        }
+
+        /// <summary>
+        /// Determine if a given SS has already been cleaned
+        /// </summary>
+        /// <param name="ss">Byte array of SS sector</param>
+        /// <returns>True if SS is clean, false otherwise</returns>
+        public static bool IsCleanSS(byte[] ss)
```

Comment:
Would all of this security sector stuff be better served in a separate library?

Reply:
I'll let you refactor these Tools to something that makes sense.. I thought they'd at least be better served outside of the XBC module as they're not specific to XBC logs. Similar to the PS3 functions above them.

Reply:
Fair enough, that can be looked at later